### ISSUE TITLE
Check and maybe add rat-human genome version

### ISSUE THREAD

**Issue by lcolladotor** [LieberInstitute/SPEAQeasy#20]

**Comment by gpertea** [comment]
Found this old issue after trying to get the `--custom_anno` option running a couple of weeks ago -- as that option should address this issue -- if that option really worked as documented.  

However that does not seem to be the case, I had to adjust some of the workflow scripts in order to sort out a few pipeline breaking errors I encountered when using that option with a custom annotation data provided in `--annotation` -- as that did not seem to work as documented in the manual.

To clarify, in my case the "custom reference and annotation" were a concatenation of human hg38 genome with Gencode 43 annotation, and Ensembl rat genome and annotation [mRatBN7.2](https://useast.ensembl.org/Rattus_norvegicus), with chromosomes prefixed with *human_* and *rat_* respectively (see `/dcs04/lieber/lcolladotor/dbDev_LIBD001/RNAseq/stemcell_pipeline/spqz_ref` on JHPCE).

The changes made to the pipeline scripts that allowed me to successfully run SPEAQeasy on such custom reference can be found in the [gjhpce_run branch](https://github.com/LieberInstitute/SPEAQeasy/tree/gjhpce_run) -- however I could not make a pull request to merge this because that branch also includes other customizations of the pipeline that are not properly documented or maybe not desirable for general use etc. so I would advise against merging that branch into master _as is_. 

However I'd be glad to discuss and help sort out the changes that could be ported to the main branch so this issue can be closed (i.e. make the `--custom_anno` option work as expected with a custom genome/annotation provided in `--annotation`)

**Comment by Nick-Eagles** [comment]
Yeah, the documentation on custom annotation is very misleading in the sense that it seems to state any annotation files work, but `--reference` must still be specified ("hg38", "hg19", "mm10", or "rn6" *only*), greatly limiting how custom things actually are (and not supporting this case, as you discovered). I imagine the bulk of the code needed to make the reference totally custom is just in this commit [here](https://github.com/LieberInstitute/SPEAQeasy/commit/7298b006ebd54f02e2cb99d425b5449ec38d595a), but I think `build_annotation_objects.R` must still be changed, and I didn't find it obvious how to totally remove species-particular hardcoded values like [here](https://github.com/LieberInstitute/SPEAQeasy/blob/2172236bb156f47b0f2259f6b08bed36036702de/scripts/build_annotation_objects.R#L41-L47). It would be great to remove hardcoded values and be as agnostic as possible to the species.

**Comment by gpertea** [comment]
There are some benefits to hard coding the Gencode species at least (human and mouse for now) as many reference/annotation features/attributes are standardized and can be uniformly managed and convenience options can be provided to the users (e.g. support for "primary" vs "main" annotation etc.). 
Beyond that, anything else could be probably considered "custom" and not hard code them, though we do have to make the specifications more clear there, e.g. the minimal set of expected attributes for the GTF annotation (`gene_name`, gene type/biotype etc.) - and encourage users to use **Ensembl** reference annotation for other species to ensure compatibility with the pipeline.

Supporting Ensembl annotations could be improved -- I noticed that the rat genome was not fully supported before my changes - in particular building the rse_tx was not possible. But it was relatively easy to correct that, specifically the transcript FASTA file can be built using gffread just based on the annotation GTF and genome sequence - and then the gene related attributes for each transcript (gene symbol, gene biotype etc.) should be pulled from the annotation GTF directly instead of the transcript FASTA headers (as reflected in the kallisto/salmon output), which is only the case for Gencode annotation.
